File: HW08/BigramDyn.java
/**
 *
 * @author furkan
 */

import java.io.*;
import java.lang.String;
import java.rmi.UnexpectedException;
import java.util.Scanner;

@SuppressWarnings("unchecked")
public class BigramDyn<T> implements Bigram<T> {
    
    private int dataType;
    private int sizeOfVal;
    private T[] mapDyn;
    
    public BigramDyn(int type){
        
        this.dataType = type;
        this.sizeOfVal=0;
        this.mapDyn = null;
    }
    
    @Override
    public  void readFile(String name) throws IOException
    {    
        Scanner input = null;
        try{
            input = new Scanner(new File(name)); 
        }
        catch(FileNotFoundException e){
            System.out.println("Dosya açılamadı !");
        } 
        
        while(input.hasNext() && !input.next().equals(" "))
            sizeOfVal++;
        
        if(sizeOfVal <=1)
            throw new UnexpectedException("Bigram olabilecek sayıda(2) yeterli"
                                                        + "  veri bulunamadı.");
                
        input.close();
        
        mapDyn = (T[]) new Object[sizeOfVal];
        input = new Scanner(new File(name)); 
        if(dataType == 1)
        {   
            Integer a,i=0;
            while(input.hasNext())
            {   
                if (input.hasNextInt()) 
                {
                    a = input.nextInt();
                    mapDyn[i]=(T)a;
                    ++i;
                }
                else
                    throw new UnexpectedException("Farklı tiplerde veri var");
                
            }
            input.close();
        }
        else if(dataType == 2)
        {
            String a;
            int i=0;
            while(input.hasNext())
            {   
                if (input.hasNext()) 
                {
                    a = input.next();
                    mapDyn[i]=(T)a;
                    ++i;
                }
                else
                    throw new UnexpectedException("Farklı tiplerde veri var");
            }
            input.close();
        }    
        else    
        {
            String text;
            int i=0;
            while(input.hasNext())
            {   
                text = input.next();
                try{
                    Double num = Double.parseDouble(text);
                    mapDyn[i]=(T)num;
                    ++i;
                }
                catch(NumberFormatException e){
                    throw new UnexpectedException("Farklı tiplerde veri var");
                }
                
            }
            input.close();
        } 
    }
    
    @Override
    public  int numGrams()
    {
        return sizeOfVal-1;
    }
    
    @Override
    public  int numOfGrams(T first,T second)
    {
        int number=0;
      
        for(int i=0;i< sizeOfVal-1;++i) 
            if((mapDyn[i].equals(first)) && (mapDyn[i+1].equals(second)))
                number++; //istenilen Bigramın kaç kez tekrarlandığı bulundu
       
        return number;
    }
    
    @Override
    public  String toString()
    {
        T arr[][]=  (T[][])new Object[sizeOfVal-1][2];
        
        for(int i=0;i< sizeOfVal-1;++i)
        {    
            arr[i][0]= mapDyn[i]; // bigramlar 2 dimensional array'de tutuldu.
            arr[i][1]= mapDyn[i+1];
        }

        int repeat[]=new int[sizeOfVal-1];

        for(int i=0; i<sizeOfVal-1;++i)
            repeat[i]= numOfGrams(arr[i][0],arr[i][1]);
            // Bigram larla aynı index te , tekrar sayıları tutuldu

        int index[]= new int[sizeOfVal-1];

        for(int i=0;i<sizeOfVal-1;++i)
            index[i]=i;

        for(int i=0;i<sizeOfVal-2;++i)    
        {                   //tekrar sayısına göre index bazlı sıralama yapıldı
            for(int j=i+1;j<sizeOfVal-1;++j)
            {
                int temp;

                if(repeat[index[i]] < repeat[index[j]])
                {
                    temp=index[i];
                    index[i]=index[j];
                    index[j]=temp;
                }    
            }    
        }    
        String answer=null;
        for(int i=0; i<sizeOfVal-1;++i)
        {
            boolean type=true;

            for(int k=0;k<i;k++)
            {   // değerin önceden print edilip edilmediğinin kontrolü 
                if( arr[index[k]][0].equals(arr[index[i]][0]) && 
                                      arr[index[k]][1].equals(arr[index[i]][1]))
                {    
                    type=false;
                    k=i;
                }
                else
                    type=true;
            }
            
            if(type)
            {    
                if(i==0)
                    answer=arr[index[i]][0]+ " , "+ arr[index[i]][1] + 
                                           "  repeat: " +repeat[index[i]]+"\n";
                else
                    answer+=arr[index[i]][0]+ " , "+ arr[index[i]][1] +
                                           "  repeat: " +repeat[index[i]]+"\n";
            }
       }
        
       return answer;
    }

} // end of BigramDyn  class

File: HW07/BigramDyn.cpp
/* 
 * File:   BigramDyn.cpp
 * Author: furkan
 * 
 * Created on 23 Aralık 2016 Cuma, 18:08
 */
#include <fstream>
#include <iostream>
#include "BigramDyn.h"
#include "exceptions.h"
using namespace std;

template <class T>
BigramDyn<T>::BigramDyn() {
    array = nullptr;
}

template <class T>
BigramDyn<T>::BigramDyn(int num):BigramDyn() {}


template <class T>
void BigramDyn<T>::readFile(const char* filename){
    
    fstream dosya;
    dosya.open(filename);
    int key =0;
    T value;
    
    if( !dosya.is_open() ) 
        throw exceptions(1); // dosya  açılamaz ise 
    else
    {
        while(!dosya.eof())
        {
            dosya >> value;
                
            if( !dosya.eof() && dosya.fail() ) // farklı tipte veriler varsa
                throw exceptions(3);
            
            key++;
        }
        key--;
    }
    if(key <=1)         // eğer 1 veya daha az değer varsa (Bigram için en az 2) 
        throw exceptions(2);
    
    size=key;
    dosya.close();
    dosya.open(filename);
    
    array = new T[key];
    for(int i=0;i<key;++i)
    {    
        dosya >> value;
        array[i]=value;
    }
    
    return;
}

template <class T>
int BigramDyn<T>::numGrams()const {
    return (size-1);        // tüm bigram sayısı
}

template <class T>
int BigramDyn<T>::numOfGrams(T first, T second){
    
    int number=0;
    
    for(int i=0;i< size;++i)
        if((i+1 != size) && (array[i] == first) && (array[i+1] == second))
            number++;       // istenilen Bigramın kaç kez tekrarlandığı bulundu
    
    return number;
}

template <class T>
ostream& BigramDyn<T>::print(ostream& out) {
    
    T **temp = new T*[size-1];
    
    for(int i=0;i<size-1;++i)
        temp[i]=new T[2];
    
    for(int i=0;i< size-1;++i)
    {    
        temp[i][0]=array[i];
        temp[i][1]=array[i+1];
    }                       // bigramlar 2 dimensional array'de tutuldu.
    
    int* repeat= new int[size-1];
    
    for(int i=0; i<size-1;++i)
        repeat[i]= numOfGrams(temp[i][0],temp[i][1]);
                // Bigram larla aynı index te , tekrar sayıları tutuldu
    
    int* index=new int[size-1];
    
    for(int i=0;i<size-1;++i)
        index[i]=i;
    
    for(int i=0;i<size-2;++i)   
    {                   //tekrar sayısına göre index bazlı sıralama yapıldı
        for(int j=i+1;j<size-1;++j)
        {
            int temp;
            
            if(repeat[index[i]] < repeat[index[j]])
            {
                temp=index[i];
                index[i]=index[j];
                index[j]=temp;
            }    
        }    
    }    
    
        
    for(int i=0; i<size-1;++i)
    {
        bool type=true;
        
        for(int k=0;k<i;k++)
        {   // değerin önceden print edilip edilmediğinin kontrolü
            if( temp[index[k]][0]==temp[index[i]][0] && 
                                           temp[index[k]][1]==temp[index[i]][1])
            {    
                type=false;
                k=i;
            }
            else
                type=true;
        }    
        if(type)
            out << temp[index[i]][0] <<" , "<<temp[index[i]][1]
                                         <<"  repeat: "<<repeat[index[i]]<<endl;
    }    
    
    for(int i=0;i<size-1;++i)
        delete [] temp[i];
        
    delete [] temp;    // alınan yerler geri verildi(gerekli değil, scope sonu)
    delete [] index;
    delete [] repeat;
    
    return out;
}

template <class T>
pair<T,T> BigramDyn<T>::maxGrams(){
    
    T value1,value2;
    value1=array[0];
    value2=array[1];
    int max = numOfGrams(value1, value2);  //ilk bigramın kaç kez tekrarlandığı
    
    for(int i=0;i+1< size;++i)
    {
        if( max< numOfGrams(array[i],array[i+1]) )
        {                                  // en çok tekrarlanan Bigram bulundu
            max = numOfGrams(array[i],array[i+1]);
            value1=array[i];
            value2=array[i+1];
        }    
    }
    pair<T,T> value;
    value= make_pair(value1,value2);
    return value;
}



template <class T>
BigramDyn<T>::~BigramDyn(){

    delete[] array;
}



